pennylane/circuit_graph.py

```diff
@@ -0,0 +1,316 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+CircuitGraph
+============
+
+**Module name:** :mod:`pennylane.circuit_graph`
+
+.. currentmodule:: pennylane.circuit_graph
+
+This module contains the CircuitGraph class which is used to generate a DAG (directed acyclic graph)
+based on operation and observable queues.
+
+.. raw:: html
+
+    <h3>Summary</h3>
+
+.. autosummary::
+    CircuitGraph
+
+.. raw:: html
+
+    <h3>Code details</h3>
+"""
+from collections import namedtuple
+import networkx as nx
+
+
+Command = namedtuple("Command", ["name", "op", "return_type", "idx"])
+Layer = namedtuple("Layer", ["op_idx", "param_idx"])
+
+
+class CircuitGraph:
+    """Represents a queue of operations and observables
+    as a directed acyclic graph.
+
+    Args:
+        queue (list[Operation]): the quantum operations to apply
+        observables (list[Observable]): the quantum observables to measure
+        parameters (dict[int->list[(int, int)]]): Specifies the free parameters
+            of the quantum circuit. The dictionary key is the parameter index.
+            The first element of the value tuple is the operation index,
+            the second the index of the parameter within the operation.
+    """
+
+    def __init__(self, queue, observables, parameters=None):
+        self._queue = queue
+        self._observables = observables
+        self.parameters = parameters or {}
+
+        self._grid = {}
+        """dict[int, list[int, Command]]: dictionary representing the quantum circuit
```

Comment:
I believe this has now changed in code itself to be `list[Command]`?

Reply:
Fixeed